### ISSUE TITLE
Getting depth texture from viewport loses precision

### ISSUE THREAD

**Issue by ratkingsminion** [godotengine/godot#85107]
### Godot version

v4.1.3.stable.official [f06b6836a]

### System information

Windows 10

### Issue description

When trying to get the depth map of a viewport, as described in the current online documentation, and using it in another shader, the linearized depth map is missing most of the depth information.

### Steps to reproduce

Create a SubViewport with Camera3D. Create two fullscreen post effect, as described in the Godot documentation (Advanced post-processing) - once for the general scene, and once for the SubViewport node. For the quad in the viewport, add a shader material with this shader (also described in the documentation):

![Screenshot_8](https://github.com/godotengine/godot/assets/3142253/bcb26f67-ddfa-4364-8c56-84c9e1168da4)

Use the viewport texture in a shader for the other post effect, linearize it (also described in the aforementioned [documentation](https://docs.godotengine.org/en/stable/tutorials/shaders/advanced_postprocessing.html)) and render it. The depth map (example is a hallway with a cylinder) should look like this (this is taken from the depth texture *not* from the viewport texture):

![Screenshot_11](https://github.com/godotengine/godot/assets/3142253/6774f041-78f1-4b84-8414-9f8ea678ed0a)

Instead it looks like this:

![Screenshot_9](https://github.com/godotengine/godot/assets/3142253/dd6895bf-918e-45fe-9710-64dc572a2cd7)

It looks like the depth texture lost precision. Interesting enough, if you activate "Use Debanding" in the SubViewport, then the result looks like this:

![Screenshot_10](https://github.com/godotengine/godot/assets/3142253/d3f9b204-9f65-415e-933c-f9dc0abd01a1)

### Minimal reproduction project

N/A

**Comment by clayjohn** [comment]
There are multiple overlapping issues here. 

Most obviously, it sounds like you are writing the depth texture information to the Viewport and then sampling that Viewport later. Doing so significantly reduces the precision of the texture as it goes through the 2D renderer (only 8 bits of precision). If you want to keep the original precision, you need to enable HDR in 2D (but that is only available in 4.2).

Second, the texture hint `source_color` might be causing an issue. It shouldn't work with the depth texture, it should just be ignored. But if it's not, it might break depth.

**Comment by ratkingsminion** [comment]
> Most obviously, it sounds like you are writing the depth texture information to the Viewport and then sampling that Viewport later. Doing so significantly reduces the precision of the texture as it goes through the 2D renderer (only 8 bits of precision). If you want to keep the original precision, you need to enable HDR in 2D (but that is only available in 4.2).

Yes, I'm doing it that way because that's how people recommend to do it online, because there's no other way to get the depth_texture from a viewport - no idea if those people tried it out or not. I know there were discussions and proposals about a way of getting a viewport's depth_texture, but I do not know why they never were implemented.

> Second, the texture hint `source_color` might be causing an issue. It shouldn't work with the depth texture, it should just be ignored. But if it's not, it might break depth.

Well, then the hint should be removed [from the official documentation](https://docs.godotengine.org/en/stable/tutorials/shaders/advanced_postprocessing.html)...

Though it doesn't change my visual output in the slightest if I remove it.

**Comment by ratkingsminion** [comment]
With Godot 4.2 RC 1 I get this result:

![Screenshot_8](https://github.com/godotengine/godot/assets/3142253/a7f5d8d0-0569-4972-8a1f-a8ad98a71b3d)

Better, but definitely still lacking. Unfortunate.

**Comment by clayjohn** [comment]
> With Godot 4.2 RC 1 I get this result:
> 
> ![Screenshot_8](https://user-images.githubusercontent.com/3142253/284161048-a7f5d8d0-0569-4972-8a1f-a8ad98a71b3d.jpg)
> 
> Better, but definitely still lacking. Unfortunate.

That is with HDR2D enabled or disabled?

**Comment by ratkingsminion** [comment]
Enabled!

**Comment by Calinou** [comment]
> Doing so significantly reduces the precision of the texture as it goes through the 2D renderer (only 8 bits of precision). If you want to keep the original precision, you need to enable HDR in 2D (but that is only available in 4.2).

Even doing that won't fully keep the precision as I understand it, as Forward+ is RGBA16F while a depth buffer is 32-bit or 24-bit.

**Comment by ratkingsminion** [comment]
Yes, that would be in line with my visual results.

As the depth_texture only uses the R component I wonder if it would be possible to 'unpack' the value into the ALBEDO's three components, and then 'pack' it later to a single value again.

**Comment by clayjohn** [comment]
> As the depth_texture only uses the R component I wonder if it would be possible to 'unpack' the value into the ALBEDO's three components, and then 'pack' it later to a single value again.

Yep, you can certainly do that. In Godot 3.x that's what we did for devices that don't support floating point textures. 

Packing code here:
https://github.com/godotengine/godot/blob/e96c42f0e162685b95da5b721701476ec5ad07eb/drivers/gles2/shaders/canvas_shadow.glsl#L53-L55

Unpack:
https://github.com/godotengine/godot/blob/e96c42f0e162685b95da5b721701476ec5ad07eb/drivers/gles2/shaders/canvas.glsl#L601

**Comment by ratkingsminion** [comment]
Thanks, I'll try it out.

**Comment by ratkingsminion** [comment]
> > As the depth_texture only uses the R component I wonder if it would be possible to 'unpack' the value into the ALBEDO's three components, and then 'pack' it later to a single value again.
> 
> Yep, you can certainly do that. In Godot 3.x that's what we did for devices that don't support floating point textures.
> 
> Packing code here:
> 
> https://github.com/godotengine/godot/blob/e96c42f0e162685b95da5b721701476ec5ad07eb/drivers/gles2/shaders/canvas_shadow.glsl#L53-L55
> 
> Unpack:
> 
> https://github.com/godotengine/godot/blob/e96c42f0e162685b95da5b721701476ec5ad07eb/drivers/gles2/shaders/canvas.glsl#L601

I tried the shader code and it doesn't work (screen is mostly black only). My guess is that the ALPHA channel of the viewport isn't interpreted correctly for this case.

**Comment by ratkingsminion** [comment]
I'm using this now to pack the depth:
```glsl
float depth = texture(depth_tex, SCREEN_UV).r;
ALBEDO = vec3(floor(depth * 2048.0), fract(depth * 2048.0), 0.0)
```

And for unpacking:
```glsl
vec4 t = texture(viewport_tex, SCREEN_UV);
float depth = t.x / 2048.0 + t.y / 2048.0;
```

This works remarkably well, and even leaves the third component free for other stuff. Still with there would be a more inbuilt solution for this.
